Add multiselect component and wire it up for country black/white lists

Lukewh opened the pull request:
Fixes https://github.com/canonical-websites/snapcraft.io/issues/1072
Fixes https://github.com/canonical-websites/snapcraft.io/issues/1074
Fixes https://github.com/CanonicalLtd/snapcraft-design/issues/490

UX: https://github.com/canonical-websites/snapcraft.io/issues/784
Design: (tags UI) https://github.com/CanonicalLtd/snapcraft-design/blob/master/snapcraft.io/market%20/Snap%20details%20-%20Listing%20view%20-%20Tags%20UI%20%20REV02.jpg

## QA

- Pull the branch
- `./run`
- Visit http://0.0.0.0:8004/account
- Click on a snap from your list of snaps
- Scroll down to the bottom of the 'Listing' page
- Change the checkbox from 'All territories' to 'Custom' under 'This snap should be available in:'
- Select either option and click on the input
- Try filtering countries by typing
- Use the mouse to scroll up and down the options
- Click on options to add them
- Use the arrow keys to scroll up and down the items
- Use the return key to add them
- Use backspace to remove your search term
- Use backspace when the search term is empty to delete the item's you've added, one by one
- Click off the element to lose focus
- Lose focus while a search term is types
- Reclick the element, the search drop-down should reappear with the term you previously searched for
- Do anything else you can think of
- ## edit fields on the Listing page to make sure that still works

# Marked as WIP until an API error is sorted - Saving works, but doesn't get served back... You can however check https://dashboard.snapcraft.io to see the changes have taken effect.

webteam-app commented:
Starting demo at: http://snapcraft.io-pr-1089.run.demo.haus/

therealjuan commented:
What about this design from [this issue](https://github.com/canonical-websites/snapcraft.io/issues/784)?

![image](https://user-images.githubusercontent.com/8167677/45498775-bc69b880-b772-11e8-899c-61c0ea774a67.png)


We can have 3 exclusive radio button options showing the relevant tag component to simplify choices.

Lukewh commented:
@therealjuan "We have our solution
A tada fireworks boom" https://github.com/canonical-websites/snapcraft.io/issues/784#issuecomment-403034504

matthewpaulthomas commented:
Two bugs:

1\. Choose “Custom list”.
2\. Type “br”.
3\. Choose “Brazil”, either with keyboard or pointer.
What should happen: The field now contains only “Brazil” (since that’s what you typed “br” to find).
What actually happens: The field contains “Brazil”, followed by the no-longer-wanted “br”.

1\. Choose “Custom list”.
2\. Type “b”.
What should happen: The list begins “Bahamas”, “Bahrain”, “Bangladesh” — that is, territories starting with what you entered. (Or maybe territories containing a word starting with what you entered, in which case the list should begin “Antigua & Barbuda”, “Bahamas”, “Bahrain”, “Bangladesh”.)
What actually happens: The list begins “Albania”, “Antigua & Barbuda”, “Aruba”, “Azerbaijan”, etc — that is, territories with a “b” _anywhere_ in their name, which is not helpful in this situation.

> I also don't think it's worth creating an 'admin page' to add four related radio buttons.

The tab is destined to contain other stuff too. So maybe it’s better to create it anyway — even if it temporarily has a more specific name, like “Territories”.

matthewpaulthomas commented on static/sass/_patterns_form-multiselect.scss:
```diff
@@ -0,0 +1,94 @@
+    align-items: flex-start;
+    border: 1px solid $color-mid-light;
+    border-radius: .125rem;
+    box-shadow: inset 0 1px 1px rgba(0, 0, 0, .12);
```
This is the same as the current `box-shadow` for Vanilla text fields. But:
- If the Vanilla style changes in future, this will be left behind.
- It gets overridden anyway: http://snapcraft.io-pr-1089.run.demo.haus/static/css/styles.css has `p-multiselect, .p-multiselect .p-multiselect__options {box-shadow: 0 1px 5px 1px rgba(17, 17, 17, 0.2); }`, making these fields inconsistent with the rest of snapcraft.io.
![snapcraft io-1089-outset](https://user-images.githubusercontent.com/19801137/45538270-7e19db00-b7fe-11e8-8b11-b7e61a5fb7ad.png)

Lukewh commented:
I've moved things to an 'Admin' page - Private, Distribution territories and metrics. I'll also add licenses to this page.

I've also fixed the bugs discovered by @matthewpaulthomas, plus a few more.

webteam-app commented:
Starting demo at: http://snapcraft.io-pr-1089.run.demo.haus/

bartaz commented on rollup.config.js:
```diff
@@ -116,4 +116,52 @@ export default [
     }
   },
+  {
+    input: 'static/js/publisher/form/multiselect.js',
```
Do we need it as separate file? This will make it bit harder when we finally migrate to webpack and also it uses react, so it means that users will download react multiple times (in each bundle that uses it).

Lukewh replied:
Good point, updated to include it as part of 'publisher.js' and included the relevant bits in the config.

bartaz commented:
### QA issues/questions

1. If I add some countries to 'Only these territories' list but then click on 'Exclude...' countries added to 'Only these' stay in place, which is confusing as the radio button is switched off, so in theory the list should be empty.

Even if we don't want to clear the value (to allow user to change their mind without loosing potentially long list of countries, we could show the multiselect as disabled to have visual clue that it's value is inactive (and won't be saved).

2. Save/Revert is on the bottom of Admin page while it's on top of Listing page.

Lukewh commented:
Renamed 'Admin' to 'Settings' as per https://github.com/CanonicalLtd/snapcraft-design/issues/490

Lukewh commented:
@bartaz I've updated the inputs so that they're disabled when unavailable. I've also moved the save/revert to the top of the page, as in listings.

webteam-app commented:
Starting demo at: http://snapcraft.io-pr-1089.run.demo.haus/

matthewpaulthomas commented:
It’s a bit unfortunate that with the pointer you can delete any territory you’ve entered, but with the keyboard you can delete only the last one. In macOS token fields, you can use arrow keys to move the insertion point between tokens, and delete either left or right. I guess this would be a lot of work to implement, because it would involve reimplementing a lot of other text field behaviour as well, rather than  just using a borderless native text field.

Simpler design issues:
- The “Metrics” controls are specifically about things that appear on the listing page, so they should remain in the “Listing” tab. (None of the designs in the linked issues suggest that they should be moved.)
- Hovering over the non-X part of a token, it darkens, suggesting that it will do something if you click it, but it doesn’t. If the tokens had a selected state, that would suggest you can type to replace the selection, but then if someone typed to replace the selection they’d expect the new text to appear in-place, which would introduce a lot of work as above. Soooo, probably better for it not to darken at all.

Bugs:

1\. Go to [/account/snaps/lukewh-test/settings](http://snapcraft.io-pr-1089.run.demo.haus/account/snaps/lukewh-test/settings)
2\. Choose “Custom list”
3\. Enter one territory, e.g. Afghanistan
4\. Look at the “Save” button, then click it
What you see: It’s greyed out and has the No cursor, but it’s clickable
What you should see: Whenever it’s clickable, it isn’t greyed out and has the normal cursor

1\. Go to [/account/snaps/lukewh-test/settings](http://snapcraft.io-pr-1089.run.demo.haus/account/snaps/lukewh-test/settings)
2\. Choose “Custom list”
3\. Enter one territory, e.g. Afghanistan
4\. Choose “Save”
What you see: “⊙ All territories”
What you should see: “⊙ Custom list” “⊙ Only these territories” “Afghanistan”

Lukewh commented:
> It’s a bit unfortunate that with the pointer you can delete any territory you’ve entered, but with the keyboard you can delete only the last one. In macOS token fields, you can use arrow keys to move the insertion point between tokens, and delete either left or right. I guess this would be a lot of work to implement, because it would involve reimplementing a lot of other text field behaviour as well, rather than just using a borderless native text field.
> 
> Simpler design issues:
> 
>     * The “Metrics” controls are specifically about things that appear on the listing page, so they should remain in the “Listing” tab. (None of the designs in the linked issues suggest that they should be moved.)
> 
>     * Hovering over the non-X part of a token, it darkens, suggesting that it will do something if you click it, but it doesn’t. If the tokens had a selected state, that would suggest you can type to replace the selection, but then if someone typed to replace the selection they’d expect the new text to appear in-place, which would introduce a lot of work as above. Soooo, probably better for it not to darken at all.
> 

I've removed the hover

> 
> Bugs:
> 
> 1. Go to [/account/snaps/lukewh-test/settings](http://snapcraft.io-pr-1089.run.demo.haus/account/snaps/lukewh-test/settings)
> 2. Choose “Custom list”
> 3. Enter one territory, e.g. Afghanistan
> 4. Look at the “Save” button, then click it
> What you see: It’s greyed out and has the No cursor, but it’s clickable
> What you should see: Whenever it’s clickable, it isn’t greyed out and has the normal cursor

This should be fixed

> 1. Go to [/account/snaps/lukewh-test/settings](http://snapcraft.io-pr-1089.run.demo.haus/account/snaps/lukewh-test/settings)
> 2. Choose “Custom list”
> 3. Enter one territory, e.g. Afghanistan
> 4. Choose “Save”
> What you see: “⊙ All territories”
> What you should see: “⊙ Custom list” “⊙ Only these territories” “Afghanistan”

As mentioned in the big text of the description of this PR:
"Marked as WIP until an API error is sorted - Saving works, but doesn't get served back... You can however check dashboard.snapcraft.io to see the changes have taken effect."

lyubomir-popov commented:
@Lukewh Is the online link updated? I don't see a checkbox All territories

bartaz commented on static/js/libs/arrays.js:
```diff
@@ -0,0 +1,23 @@
+  oldArray.sort();
+  newArray.sort();
+
+  for(let i = 0; i < oldArray.length; i++) {
```
Is `sort` needed? If arrays are meant to be equal can't we just check elements one by one without sorting? Or does `arraysEqual` checks not for equality but just if arrays contain exactly the same elements (in any order?) - in which case the name of the function is confusing.

Lukewh replied:
The sorting is a catch-all in case there's some sorting elsewhere

bartaz commented on templates/publisher/settings.html:
```diff
@@ -0,0 +1,220 @@
+        .getElementById('settings-form')
+        .addEventListener(
+          'change',
+          snapcraft.publisher.settings.changeHandler
```
Why this is separate from `initForm`? Does it handle different things that the form does? If this is not a market page why it uses `market.initForm`?

Lukewh replied:
I'm tempted to rename market.js to form.js... I wanted to reuse all the state logic you'd already written so I did, though it's true, this is not the market page.

bartaz replied:
If it's shared makes sense to rename it - would be less confusing.

Lukewh replied:
Renamed.

The reason this is here is that is shows/hides the custom radio's based on what's selected. I didn't want to add yet more to the marketForm js.

I agree this should be refactored at a later date, now we have 2 separate forms using the same code.... Probably a good time to reactify it too.

Lukewh commented on static/js/libs/arrays.js:
```diff
@@ -0,0 +1,23 @@
+    return false;
+  }
+
+  oldArray.sort();
```
Yeah so before the sorting the function checks "are they the same array?" if not is either of them null, or are the lengths different? Then is sorts both of them and checks each element in the array, the side effects or sorting shouldn't matter as after the sort it's checking on an item by item basis.

bartaz replied:
By side effects I mean you are chaning the arrays you are 'checking'.

```js
a = [3, 2, 1];
b = [2, 1, 3];
arraysEqual(a, b); // true
a; // [1, 2, 3]
b; // [1, 2, 3]
```

bartaz replied:
Unless this is expected and you want to make the arrays equal?

Lukewh replied:
Oh I see! I'll add a param to the function to indicate whether you want sorting or not

bartaz replied:
@Lukewh if `arraysEquals` is meant only to check if 2 arrays contain same elements (and from its usage it seems it is) it should never change the arrays given as params. If sorting is required to allow checking for items in any order it should be done on clones of original arrays, so that objects given to the function would not be affected by it.

Lukewh replied:
of course, sorry baby brain.

bartaz commented on static/js/publisher/form/multiselect.js:
```diff
@@ -0,0 +1,376 @@
+        case 'Backspace':
+          if (this.state.selected.length > 0 && this.state.searchTerm === '') {
+            this.removeItem({
+              target: {
```
This feels a bit too hacky - building fake event object with just the fields `removeItem` uses...
Please create separate `removeItemClick(event)` and `removeItem(whatever, params, it, needs)`.

bartaz commented on static/js/publisher/market/state.js:
```diff
@@ -7,9 +10,18 @@ const allowedKeys = [
   'private',
   'public_metrics_enabled',
-  'public_metrics_blacklist'
+  'public_metrics_blacklist',
```
If this file starts to be reused in different places I feel it may require some refactor.

It was developed for market page form and has a hardcoded list of market page inputs. Now we added other inputs here that we moved to a different page. So now we have a 'generic' form state library that contains list of fields from 2 different forms.

Maybe `state` could stay generic but it would receive `allowedKeys` and `transform` functions as parameters when created/initialized? This way shared functionality would stay here while form dependant list of the fields would be separated.

Lukewh replied:
See above re. refactor

bartaz commented on static/js/publisher/form/multiselect.js:
```diff
@@ -0,0 +1,376 @@
+   * Update the original input value and dispatch a change event to the input and form.
+   */
+  updateInputValue() {
+    this.props.input.setAttribute('value', this.state.selected.map(item => item.key).join(', '));
```
Wouldn't `this.props.input.value  = ...` do the same thing? And additionaly it should also trigger change event. There really shouldn't be a need to trigger change event manually on a form.

Lukewh replied:
aah the magical `new Event('change', { 'bubbles': true });` :+1: thanks

bartaz commented on static/js/publisher/form/multiselect.js:
```diff
@@ -0,0 +1,376 @@
+          event.preventDefault();
+          if (highlighted >= 0) {
+            this.addItem({
+              target: {
```
See the comment about `removeItem`. Please separate `addItem(key)` and `addItemClick(event)` functions not to duct tape fake event object.

Lukewh commented on static/js/publisher/settings.js:
```diff
@@ -0,0 +1,40 @@
+function enableInput(type) {
```
I'll create a new issue for this... This PR has gone wildly out of control considering it was to add a simple blacklist/ whitelist functionality and we've ended up with an entirely new page.

Lukewh replied:
https://github.com/canonical-websites/snapcraft.io/issues/1125

bartaz commented on static/js/libs/arrays.js:
```diff
@@ -0,0 +1,26 @@
+function arraysEqual(oldArray, newArray) {
```
Just add a comment explaining what the function does (especially to mention that arrays containing same elements in different order are considered equal).

Lukewh replied:
Done

bartaz commented on static/js/publisher/form.js:
```diff
@@ -1,6 +1,7 @@
-import { initSnapScreenshotsEdit } from './screenshots';
+import { initSnapScreenshotsEdit } from './market/screenshots';
```
This file got quite confusing now. While it has some bits that are common to both forms it has also a lot of things for one or another - like unnecessary screenshots on admin page or unnecessary whitelist multiselect on market page.

But let's handle it in the refactor task.

bartaz commented on static/js/publisher/form/multiselect.js:
```diff
@@ -0,0 +1,355 @@
+    super(props);
+
+    // hide the original input
+    this.props.input.style.display = 'none';
```
Ideally I would say that React components should not touch DOM directly by themselves. Things like that would be better off in some 'init' script that would read the data from input, hide it, and pass the data into React component.

The `init` function down below this class seems like good place to do it.

If it complicates things too much now I'm fine with leaving it for refactor PR which should clear up some things.

Lukewh replied:
Moved!

webteam-app commented:
Starting demo at: http://snapcraft.io-pr-1089.run.demo.haus/

bartaz commented:
### QA

It seems that country dropdown is not focused/selected by default when the page first loads which makes it harder to enter something with keyboard only.

To reproduce:
1. go to settings page of a snap that doesn't have any custom countries set
2. click on 'custom' checkbox
3. start typing contry name
4. dropdown shows up but no country is selected, so even if only one country is in dropdown it cannot be added with `enter`
5. you have to push arrow down to focus/select show up in the dropdown

<img width="1024" alt="screen shot 2018-09-24 at 14 08 36" src="https://user-images.githubusercontent.com/83575/45951452-ce194e80-c003-11e8-985b-1a5627bf7d0b.png">

It would be good to have one value in dropdown selected by default. It seems to work this way after one item gets manually selected.

bartaz commented:
### QA

Tabbing out of multiselect dropdown doesn't make it look like it looses focus.

1. Go into multiselect dropdown
2. start typing
3. hit `Tab`
4. multiselect looses focus, but it still looks focused

<img width="1033" alt="screen shot 2018-09-24 at 14 23 31" src="https://user-images.githubusercontent.com/83575/45951932-a925db00-c005-11e8-8602-9622f4816d19.png">

If multiselect looses focus (either by mouse or keyboard) it should visually loose focus as well.

bartaz commented:
### QA

Can't get back to 'All teritories'.

1. Select some territories in 'Only these countries'
2. Save
3. Click 'All teritories'
4. Save
5. It says that changes were saved correctly, but nothing changes

<img width="1015" alt="screen shot 2018-09-24 at 14 33 46" src="https://user-images.githubusercontent.com/83575/45952308-dfb02580-c006-11e8-8a39-b9115950d9f6.png">

bartaz commented:
### QA

Can't change from 'Only' territories to 'Exclude' territories

1. Add some territories to 'only' list
2. Save
3. Switch to 'Exclude' territories and add some
4. Save
5. Key 'extra' error

<img width="1018" alt="screen shot 2018-09-24 at 14 34 45" src="https://user-images.githubusercontent.com/83575/45952381-2271fd80-c007-11e8-8de7-cea2deb102dc.png">

webteam-app commented:
Starting demo at: http://snapcraft.io-pr-1089.run.demo.haus/

bartaz commented:
### QA

There is no snap name in browser page title:

<img width="446" alt="screen shot 2018-09-25 at 15 31 55" src="https://user-images.githubusercontent.com/83575/46017610-35a0ce00-c0d8-11e8-8c88-0187ba373428.png">

bartaz commented:
### QA

KeyError: 'extra' when trying to change from 'Only these' to 'Exclude' with keyboard only.

- go to settings page
- set some 'only these territories'
- Save
- Only using keyboard tab into territories radio button, change to 'Exclude' territories
- Type some territories
- Shift-Tab to 'Save' button
- Save
- You get a key error, nothing is saved

<img width="1042" alt="screen shot 2018-09-25 at 15 36 22" src="https://user-images.githubusercontent.com/83575/46017976-1191bc80-c0d9-11e8-91ac-0f4fa1f05878.png">

webteam-app commented:
Starting demo at: http://snapcraft.io-canonical-websites-pr-1089.run.demo.haus/

therealjuan commented:
- Can we change `Custom list` to `Selected territories` ?
- Can we include some padding / margin between `All territories | Selected territories` and likewise between `[ ] Only for...` and `[ ] Excluse these...` with the super select tag component?
- Would it be much work to include a `Clear list` action on the top right of the text field with more than 2 or 3 values` [ ] Only these territories` (left) `Clear list` (right)?

*Note:* There is a territory called [Christmas Island](https://en.wikipedia.org/wiki/Christmas_Island)

You can go quite far with the component...
![image](https://user-images.githubusercontent.com/8167677/46096536-5f7cf200-c1b7-11e8-95f5-162d14917fe7.png)

webteam-app commented:
Starting demo at: http://snapcraft.io-canonical-websites-pr-1089.run.demo.haus/

webteam-app commented:
Starting demo at: http://snapcraft.io-canonical-websites-pr-1089.run.demo.haus/

webteam-app commented:
Starting demo at: http://snapcraft.io-canonical-websites-pr-1089.run.demo.haus/

webteam-app commented:
Starting demo at: http://snapcraft.io-canonical-websites-pr-1089.run.demo.haus/

therealjuan commented:
LGTM
